Source/WebCore/css/CSSValue.h

```diff
@@ -154,195 +162,319 @@ class CSSValue {
-        FontPaletteValuesOverrideColorsClass,
-        FontStyleRangeClass,
-        FontStyleWithAngleClass,
-        FontVariantAlternatesClass,
-        FontVariationClass,
-        GridLineNamesClass,
-        GridTemplateAreasClass,
-        InsetShapeClass,
-        LineBoxContainClass,
-        OffsetRotateClass,
-        PathClass,
-        PendingSubstitutionValueClass,
-        QuadClass,
-        RayClass,
-        RectClass,
-        ReflectClass,
-        ShadowClass,
-        UnicodeRangeClass,
-        ValuePairClass,
-        VariableReferenceClass,
+        PrefixedLinearGradient,
+        PrefixedRadialGradient,
+        RadialGradient,
+        // Update isInImageFamily if the last image generator class changes.
 
         // Classes that contain vectors, which derive from CSSValueContainingVector.
-        ValueListClass,
-        FunctionClass,
-        GridAutoRepeatClass,
-        GridIntegerRepeatClass,
-        ImageSetClass,
-        PolygonClass,
-        SubgridClass,
-        TransformListClass,
+        // Keep ValueList as the first of these.
+        ValueList,
+        Function,
+        GridAutoRepeat,
+        GridIntegerRepeat,
+        Polygon,
+        Subgrid,
+        TransformList,
         // Do not append classes here unless they derive from CSSValueContainingVector.
     };
+    static constexpr auto LastType = Type::TransformList;
 
-    constexpr ClassType classType() const { return static_cast<ClassType>(m_classType); }
-
-    explicit CSSValue(ClassType classType)
-        : m_classType(classType)
-    {
-    }
+    explicit CSSValue(Type);
+    Type type() const;
 
-    void makeStatic()
-    {
-        m_refCount |= refCountFlagIsStatic;
-    }
-
-    // NOTE: This class is non-virtual for memory and performance reasons.
-    // Don't go making it virtual again unless you know exactly what you're doing!
+    // This class does not use virtual functions for memory and performance reasons.
     ~CSSValue() = default;
     WEBCORE_EXPORT void operator delete(CSSValue*, std::destroying_delete_t);
 
-    ValueSeparator separator() const { return static_cast<ValueSeparator>(m_valueSeparator); }
+    ValueSeparator separator() const;
     static ASCIILiteral separatorCSSText(ValueSeparator);
     ASCIILiteral separatorCSSText() const { return separatorCSSText(separator()); };
 
 private:
+    // Encode scalar values in the pointer.
+    static constexpr uintptr_t HasScalarBit = 1;
+    static constexpr uint8_t TypeShift = 1;
+    static constexpr uint8_t TypeBits = 7;
+    static constexpr uintptr_t TypeMask = (1 << TypeBits) - 1;
+
+protected:
+    uintptr_t uncheckedScalar() const { return opaque(reinterpret_cast<uintptr_t>(this)); }
+    bool hasScalarInPointer() const { return uncheckedScalar() & HasScalarBit; }
+    static bool eitherHasScalarInPointer(const CSSValue&, const CSSValue&);
+    uintptr_t scalar() const { ASSERT(hasScalarInPointer()); return uncheckedScalar(); }
```

Comment:
It is bit unclear to me why this is so focused (in terms of naming) on integer "scalars" but I suppose things can be generalised later if needed.

Reply:
I’d be happy to switch to another naming scheme; this was just the best I could come up with. I use “scalar” to mean “non-pointer encoded in the bits that would normally be a pointer”, not to mean that the values it encodes are integral.